message_ix/tests/test_feature_temporal_level.py

```diff
@@ -0,0 +1,568 @@
+
+
+def test_two_seasons_to_year_relative(test_mp):
+    comment = "3a.linked-two-seasons-with-year-relative"
+    # Dictionary of technology input/output
+    tec_dict = {
+        "gas_ppl": {
+            "time_origin": ["year", "year"],
+            "time": ["summer", "winter"],
+            "time_dest": ["summer", "winter"],
+        },
+        "gas_supply": {"time_origin": [], "time": ["year"], "time_dest": ["year"]},
+    }
+    # Shouldn't pass the test, if adding relative duration time
+    try:
+        model_generator(
+            test_mp,
+            comment,
+            tec_dict,
+            time_steps=[
+                ("summer", 0.5, "season", "year"),
+                ("winter", 0.5, "season", "year"),
+            ],
+            relative_time=["year"],
+            demand={"summer": 1, "winter": 1},
+        )
+    except AssertionError:
+        pass
+
+
+# 3b) "gas_supply" and "gas_ppl" linked at each season
+# Results: Model should solve and the linkage is made ("gas_supply" is active)
+def test_seasons_to_seasons(test_mp):
+    comment = "3b.linked-two-seasons-with-two-seasons"
+    # Dictionary of technology input/output
+    tec_dict = {
+        "gas_ppl": {
+            "time_origin": ["summer", "winter"],
+            "time": ["summer", "winter"],
+            "time_dest": ["summer", "winter"],
+        },
+        "gas_supply": {
+            "time_origin": [],
+            "time": ["summer", "winter"],
+            "time_dest": ["summer", "winter"],
+        },
+    }
+
+    model_generator(
+        test_mp,
+        comment,
+        tec_dict,
+        time_steps=[
+            ("summer", 0.5, "season", "year"),
+            ("winter", 0.5, "season", "year"),
+        ],
+        demand={"summer": 1, "winter": 1},
+    )
+
+
+# 4) Meeting "demand" through three temporal levels ("month", "season", and "year")
+# 4a) "month" is defined under "season" BUT "season" not linked to "year"
+# Results: Model should not solve (no parent "time" for "season", i.e., no
+# linkage to "gas_supply" at "year")
+def test_unlinked_three_temporal_levels(test_mp):
+    comment = "4a.unlinked-temporal-levels"
+    # Dictionary of technology input/output
+    tec_dict = {
+        "gas_ppl": {
+            "time_origin": ["year", "year"],
+            "time": ["Jan", "Feb"],
+            "time_dest": ["Jan", "Feb"],
+        },
+        "gas_supply": {
+            "time_origin": [],
+            "time": ["year"],
+            "time_dest": ["year"],
+        },
+    }
+
+    # Check the model shouldn't solve
+    try:
+        model_generator(
+            test_mp,
+            comment,
+            tec_dict,
+            time_steps=[
+                ("Jan", 0.25, "month", "winter"),
+                ("Feb", 0.25, "month", "winter"),
+                ("Jun", 0.25, "month", "summer"),
+                ("Jul", 0.25, "month", "summer"),
+            ],
+            demand={"Jan": 1, "Feb": 1},
+        ),
+    except ModelError:
```

Comment:
This is a good test! Looking for specific errors. The idiomatic way to do this in pytest is described [here](https://docs.pytest.org/en/stable/getting-started.html#assert-that-a-certain-exception-is-raised):
```py
with pytest.raises(ModelError):
    model_generator(...)
```

Reply:
Nice. Implemented.